pkg/metrics/metrics.go

```diff
@@ -126,27 +192,91 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
 		opMgr.cache.Delete(op)
 		return
 	}
-	strStatus := opStatusUnknown
+	strStatus := string(SnapshotStatusTypeUnknown)
 	if status != nil {
 		strStatus = status.String()
 	}
-	duration := time.Since(ts).Seconds()
-	opMgr.opLatencyMetrics.WithLabelValues(op.Driver, op.Name, strStatus).Observe(duration)
+
+	// record invidual reconciliation time
+	reconciliationDuration := time.Since(ts).Seconds()
+	opMgr.opReconciliationLatencyMetrics.WithLabelValues(op.Driver, op.Name, op.SnapshotType, strStatus).Observe(reconciliationDuration)
+
+	// record total operation durations if operation is successful
+	if strStatus == string(SnapshotStatusTypeSuccess) {
+		switch op.Name {
+		case CreateSnapshotOperationName:
+			// time since user first created volumesnapshot until a volumesnapshot.Status.CreationTime is aded
+			operationDuration := time.Since(snapshot.CreationTimestamp.Time).Seconds()
```

Comment:
it should be: `time.Since(cached-start-time)`. the creation timestamp of snapshot object does not mean the time when the controller sees it. In an extreme case, say when the snapshot controller has not been installed initiatively, the user creates a snapshot object, and x-hours later the snapshot controller got installed, we should *only* start counting when the controller gets to process the snapshot object.

Reply:
I disagree here. I think the end to end metrics should reflect the user experience, which includes time spent in other components. If there's a bug in the snapshot controller such that it restarts frequently or is slow to process a snapshot request, we should capture that.

Reply:
I agree slowness should be captured, but the point is not related here, it will be captured actually.
Crash-looping is a different case though. If controller is crash looping, the user should have been notified by other manner not the e2e metrics.
The point is that the controller is not installed at all when a snapshot CR is created, in this case, using the creation time stamp of a snapshot CR as the start time is not reasonable to me. it will create a weird spiky point in the time-series, and I suspect this is the experience a user would want.
setting up a meeting to discuss further tomorrow.

Reply:
How is this kind of metric calculated for PVC creation?  Do we use creation timestamp of PVC as the the start time or do we use the time when the PV controller sees the PVC object?

Reply:
I think there's value to have multiple metrics each capturing different levels of response, starting with the user experience and getting more fine grained to the controller, then the csi driver.

The higher level metrics can be used to define SLOs, whereas the finer grained metrics can be used to debug where something may be going wrong.

Reply:
Sure.  I think we should keep this metric and modify the description to reflect what it measures.
As a followup, add another metric that calculates the time between the controller observed start time and readyToUse.